onnx/version_converter/adapters/axis_input_to_attribute.h

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) ONNX Project Contributors
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "onnx/version_converter/adapters/adapter.h"
+
+namespace ONNX_NAMESPACE {
+namespace version_conversion {
+class AxisInputToAttribute : public Adapter {
+ public:
+  explicit AxisInputToAttribute(
+      const std::string& op_name,
+      const OpSetID& initial,
+      const OpSetID& target,
+      size_t axis_index,
+      int64_t default_axis)
+      : Adapter(op_name, initial, target), axis_index(axis_index), default_axis(default_axis) {}
+
+  Node* adapt(std::shared_ptr<Graph> graph, Node* node) const override {
+    if (!HasAxisInput(node)) {
+      node->i_(kaxis, this->default_axis);
+      return EnsureAndReturnNode(node);
+    }
+
+    const ArrayRef<Value*>& inputs = node->inputs();
+    Value* axis_val = inputs[this->axis_index];
+    Node* axis_node = axis_val->node();
+
+    if (axis_node->kind() == kConstant) {
+      HandleConstantNode(node, axis_node, axis_val);
+      return EnsureAndReturnNode(node);
+    }
+
+    HandleInitializerNode(graph, node, inputs, axis_val);
+    return EnsureAndReturnNode(node);
+  }
+
+ private:
+  size_t axis_index;
+  int64_t default_axis;
+
+  bool HasAxisInput(const Node* node) const {
+    const ArrayRef<const Value*>& inputs = node->inputs();
+    return inputs.size() > this->axis_index && inputs[this->axis_index]->node()->kind() != kUndefined;
+  }
+
+  void HandleConstantNode(Node* node, Node* axis_node, Value* axis_val) const {
+    const std::vector<int64_t>& int64s = axis_node->t(kvalue).int64s();
+    if (int64s.empty()) {
+      std::string raw_data = axis_node->t(kvalue).raw();
+      ONNX_ASSERTM(
+          raw_data.size() != 0 && raw_data.size() % 8 == 0,
+          "Raw Data must be non-empty and size must be a multiple of 8");
+      const int64_t* raw = reinterpret_cast<const int64_t*>(raw_data.c_str());
+      node->i_(kaxis, raw[0]);
+    } else {
+      node->i_(kaxis, int64s.at(0));
+    }
+    node->removeInput(this->axis_index);
+    if (axis_val->uses().size() < 1) {
+      node->destroy();
+    }
+  }
+
+  void HandleInitializerNode(std::shared_ptr<Graph> graph, Node* node, const ArrayRef<Value*>& inputs, Value* axis_val)
+      const {
+    for (const auto& initializer : graph->initializers()) {
+      if (initializer.name() == inputs[this->axis_index]->uniqueName()) {
+        node->i_(kaxis, initializer.int64s().at(0));
+        node->removeInput(this->axis_index);
+        // Remove initializer
+        if (axis_val->uses().size() < 1)
+          graph->eraseInitializerAndInput(axis_val);
```

Comment:
This looks incorrect. An initializer should be treated as a constant only if it is NOT also an input. (If this pattern is from somewhere else, it is possible that code may be incorrect as well.) If it is also an input, it should be handled as an unknown axis value (which means, the conversion should fail).

Reply:
I copied it. Will update. Thanks!